rfcs/rfc-story-based-git-workflow.md

```diff
@@ -0,0 +1,117 @@
+
+Adding team Bumblebee (FreshWorks) as codebase contributors to the <https://github.com/bcgov/lear> repository exposed some workflow issues around being able to continuously develop with non co-located teams comprised of many team members. The review and commit strategies that were effective with one team do not scale to two teams. There are some other advantages to story branching that can be realized as well.
+
+Here are some of the issues we have seen:
+
+* Frequent small commits that invalidate the test cycle for a story that is almost ready to be created as a pull request
+* Lack of transparency between developers' work in progress (individual developer forks live on GitHub accounts on an ad-hoc basis)
+* Commits to master that are incomplete or break the build
+* Commits to master that do not pass linting
+* Developers committing their own code without a review
+
+# Detailed design
+
+No detailed design necessary as the actual technical aspects are not complicated. This RFC describes a potential team agreement for WoW (way of working), not a technical design change.
+
+# Drawbacks
+
+- Disrupting the current team's WoW (Way of Working)
+- Not consistent with how public committers would be working (if we were using them - although we could create branches for them as well if the volume is low)
+- Currently difficult to do E2E testing in local development environment
+- Risk of "stories" being defined in a very course-grained way and loooooong running PRs being difficult/impossible to merge
+
+
+# Alternatives
+
+## Running tests in a PR environment before merging
+If an E2E environment can be set up in an automated fashion, webhooks can be used to trigger a dedicated environment set up when the PR is created. All the automated tests will be run in that environment. 
+
+**Advantages of PR environment:**
+ 
+* Automated tests are scripted to run in OpenShift, not local (less chance of "works on my machine")
+* PR reviewer does not have to clone the branch and run the tests themselves -- they can just focus on the code and trust that it works.
+* There is an environment where the working story or story can be demoed or manually tested
+* Reduced risk of incomplete or buggy code being merged into master -- shared program E2E environment will be more stable
+
+**Disadvantages of PR environment:** 
+
+* Consumes OpenShift resources
+* Requires DevOps effort (current build/deploy/test cycle is not fully scripted)
+* Pipelines can be slow
+
+**Recommendation**: Work toward automated, scripted, consistent environments
+
+## Forking instead of Branching
+To date, the entities team has not been using branches. Because the code is open source, developers can fork and share their fork between each other on an ad-hoc basis. 
```

Comment:
Incorrect, all work is done on branches. Please refer to how the work is actually done.

Reply:
Can you please clarify? My understanding is that everyone forks. Are you saying that the forks are then branched and the work is done on the forked branch? Not sure the difference between working on the fork or the branch in that sense. Does that make it easier to rebase (because you can rebase your forked master without affecting your branch)? Please clarify for my sake.

Reply:
OK, I found this: https://github.com/bcgov/namex/blob/master/docs/developer.md 
Now I understand what you were trying to say. Will change to say "has not been using branches directly off master in the same repo"